### PR TITLE
Shell escape

### PR DESCRIPTION
**Pull request by creativeprojects**
The current implementation of passing (funny) parameters to restic is broken, as explained in #56 

Let's take this config as an example:

```yaml
---
default:
  password-file: key
  repository: "rest:http://user:password@localhost:8000/path"

simple:
  inherit: default
  backup:
    exclude: "/**/.git"
    source: /source

spaces:
  inherit: default
  backup:
    exclude: "My Documents"
    source: /source dir

glob:
  inherit: default
  backup:
    exclude:
      - "examples/integration*"
    source:
      - "examples/integration*"

quote:
  inherit: default
  backup:
    source:
      - "/Côte d'Ivoire"
```

For each profile, here are the differences between the current implementation and the fix:

## simple

This is a simple configuration with no funny character (no glob, no space, just plain). **This is what most people will have as a configuration.**

`resticprofile -n spaces backup --option`

Current and new implementations are sending the same arguments, there's no change. Here's the list of arguments restic will receive:

`backup`
`--exclude`
`/**/.git`
`--password-file`
`key`
`--repo`
`rest:http://user:password@localhost:8000/path`
`--option`
`/source`

## spaces

Current implementation fails if you add some parameters containing spaces on the command line:

`resticprofile -n spaces backup "some path"`

Current | Fix
--- | ---
`backup` | `backup`
`--exclude` | `--exclude`
`My Documents` | `My Documents`
`--password-file` | `--password-file`
`key` | `key`
`--repo` | `--repo`
`rest:http://user:password@localhost:8000/path` | `rest:http://user:password@localhost:8000/path`
`some` | `some path`
`path` |
`/source dir` | `/source dir`

## glob

`resticprofile -n glob backup examples/integration*`

Let's imagine we have 2 files in `examples/` named `integration_test.toml` and `integration_test.yaml`

Current implementation is working fine in this case, no change.

`backup`
`--exclude`
`examples/integration*`
`--password-file`
`key`
`--repo`
`rest:http://user:password@localhost:8000/path`
`examples/integration_test.toml`
`examples/integration_test.yaml`
`examples/integration_test.toml`
`examples/integration_test.yaml`

## quote

`resticprofile -n quote backup "/path/with space; echo foo"`

Adding a single quote in a name actually breaks the current implementation with the shell returning an error 2 (there's no closing quote)

Current | Fix
--- | ---
`backup` | `backup`
`--password-file` | `--password-file`
`key` | `key`
`--repo` | `--repo`
`rest:http://user:password@localhost:8000/path` | `rest:http://user:password@localhost:8000/path`
`/path/with` | `/path/with space; echo foo`
`space;` |
`echo` |
`foo` |
`/Côte d'Ivoire` | `/Côte d'Ivoire`

**Unless anybody objects to this new implementation, I will merge the fix soon-ish** 👍🏻

### CODE REVIEW

**Comment by jkellerer** [comment]
Regarding the `source` list defined in a profile: I think this should always be escaped and as a consequence this requires to resolve glob expressions internally using `filepath.Glob()` instead of relying on shell expansion.

Consider the following example:
```yaml
----
profglob:
  repository: rest:https://user:pass@remotehost
  password-file: test-repo.key

  retention:
    after-backup: true
    keep-last: 3

  backup:
    source: path*
```

This results in the following command lines when running `./resticprofile -n profglob --dry-run backup`:
````
/usr/local/bin/restic backup --password-file test-repo.key --repo rest:https://user:***@remotehost path*
/usr/local/bin/restic forget --keep-last 3 --password-file test-repo.key --path /test/path* --repo rest:https://user:***
````

For `backup` this works fine, but in `forget` it breaks, since all paths must be written as "--path path1 --path path2 --path pathX ...".

Note: It may also make sense (but this is out of scope for this PR) to introduce `path` to `retention` which can be set to `false` (not add a path filter), `true` (current behaviour) and list of paths overriding `source`. In general it makes sense that `retention` filters by paths but for glob expressions it may not since paths are not stable (dynamic) and snapshots may not match.

File: shell/arg.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	doubleQuotePattern = regexp.MustCompile(`[^\\][|&;<>()$'" \t\r\n*?[]`)
+)
+
+type ArgType int
+
+const ArgTypeCount = 4
+
+const (
+	ArgConfigEscape       ArgType = iota // escape each special character but don't add quotes
+	ArgConfigNoGlobQuote                 // use double quotes around argument when needed
```
Comment (jkellerer) [inline shell/arg.go:22]: The name NoGlob (and "no-glob") made me think at first that glob expressions are forbidden, but in fact it means that glob expressions are passed to restic (as required for include/exclude patterns).

I know this is just internal, but would it make sense to use a name like "preserve-glob" or "keep-glob"?

Comment (creativeprojects) [reply shell/arg.go:22]: You're right, that helps with code readability 😛

**Comment by creativeprojects** [comment]
> 
> For `backup` this works fine, but in `forget` it breaks, since all paths must be written as "--path path1 --path path2 --path pathX ...".
> 

You see, it's funny because I knew I was forgetting something but couldn't figure out what, hence why I was leaving the PR opened 🤔

**Pull request merged**